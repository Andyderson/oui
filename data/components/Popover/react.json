{"description":"Simple component often used to display supplemental information to contents\non a page.\n\nSee `OverlayWrapper` component to position a popover on a page.","methods":[],"props":{"children":{"type":{"name":"node"},"required":true,"description":"Content that appears within the popover body"},"testSection":{"type":{"name":"string"},"required":false,"description":"Hook for automated JavaScript tests"},"title":{"type":{"name":"string"},"required":false,"description":"Text describing the popover contents"}},"private":false}