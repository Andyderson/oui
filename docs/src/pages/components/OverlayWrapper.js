
  import React from 'react';
  import ComponentContent from '../../layouts/templates/ComponentContent';

   import data from '../../../../data/components/OverlayWrapper/react.json'; import examples from '../../../../src/components/OverlayWrapper/example'; import readme from 'raw-loader!../../../../src/components/OverlayWrapper/README.md';

  let dataObject = {
    react: data ? data : null, 
    examples: examples ? examples : null,
    readme: readme ? readme : null
  }
  
  class OverlayWrapperComponent extends React.Component {
  
    render() {
      return (
        <ComponentContent data={ dataObject } />
      );
    };
  }
  
  export default OverlayWrapperComponent
