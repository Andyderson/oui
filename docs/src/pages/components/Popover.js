
  import React from 'react';
  import ComponentContent from '../../layouts/templates/ComponentContent';

   import data from '../../../../data/components/Popover/react.json'; import examples from '../../../../src/components/Popover/example'; import readme from 'raw-loader!../../../../src/components/Popover/README.md';

  let dataObject = {
    react: data ? data : null, 
    examples: examples ? examples : null,
    readme: readme ? readme : null
  }
  
  class PopoverComponent extends React.Component {
  
    render() {
      return (
        <ComponentContent data={ dataObject } />
      );
    };
  }
  
  export default PopoverComponent
