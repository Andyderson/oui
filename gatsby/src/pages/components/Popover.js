
  import React from 'react';
  import ComponentContent from '../../layouts/templates/ComponentContent';

   import examples from '../../../../src/components/Popover/example'; import readme from 'raw-loader!../../../../src/components/Popover/README.md';

  const sassData = {"name":"Popover\n","description":"Information bubbles on hover.","start":{"line":1,"column":1},"end":{"line":112,"column":4},"example":[{"type":"html","description":null,"code":"<div class=\"height--100\">\n  <ul class=\"oui-pop--over oui-pop--over--arrow-top-left\" style=\"display: block; opacity: 1;\">\n    <li class=\"oui-pop--over__content\">\n      <div class=\"oui-pop--over__title\">Title</div>\n      <p>Ipsa officiis bad-news minus earum a aperiam! Aperiam reiciendis vitae nihil libero et, hic!</p>\n    </li>\n  </ul>\n</div>"},{"type":"html","description":null,"code":"<div class=\"height--100\">\n  <ul class=\"oui-pop--over oui-pop--over--arrow-top-center\" style=\"display: block; opacity: 1;\">\n    <li class=\"oui-pop--over__content\">\n      <div class=\"oui-pop--over__title\">Title</div>\n      <p>Ipsa officiis bad-news minus earum a aperiam! Aperiam reiciendis vitae nihil libero et, hic!</p>\n    </li>\n  </ul>\n</div>"},{"type":"html","description":null,"code":"<div class=\"height--100\">\n  <ul class=\"oui-pop--over oui-pop--over--arrow-top-right\" style=\"display: block; opacity: 1;\">\n    <li class=\"oui-pop--over__content\">\n      <div class=\"oui-pop--over__title\">Title</div>\n      <p>Ipsa officiis bad-news minus earum a aperiam! Aperiam reiciendis vitae nihil libero et, hic!</p>\n    </li>\n  </ul>\n</div>"},{"type":"html","description":null,"code":"<div class=\"height--100\">\n  <ul class=\"oui-pop--over oui-pop--over--arrow-right-top\" style=\"display: block; opacity: 1;\">\n    <li class=\"oui-pop--over__content\">\n      <div class=\"oui-pop--over__title\">Title</div>\n      <p>Ipsa officiis bad-news minus earum a aperiam! Aperiam reiciendis vitae nihil libero et, hic!</p>\n    </li>\n  </ul>\n</div>"},{"type":"html","description":null,"code":"<div class=\"height--100\">\n  <ul class=\"oui-pop--over oui-pop--over--arrow-right-center\" style=\"display: block; opacity: 1;\">\n    <li class=\"oui-pop--over__content\">\n      <div class=\"oui-pop--over__title\">Title</div>\n      <p>Ipsa officiis bad-news minus earum a aperiam! Aperiam reiciendis vitae nihil libero et, hic!</p>\n    </li>\n  </ul>\n</div>"},{"type":"html","description":null,"code":"<div class=\"height--100\">\n  <ul class=\"oui-pop--over oui-pop--over--arrow-right-bottom\" style=\"display: block; opacity: 1;\">\n    <li class=\"oui-pop--over__content\">\n      <div class=\"oui-pop--over__title\">Title</div>\n      <p>Ipsa officiis bad-news minus earum a aperiam! Aperiam reiciendis vitae nihil libero et, hic!</p>\n    </li>\n  </ul>\n</div>"},{"type":"html","description":null,"code":"<div class=\"height--100\">\n  <ul class=\"oui-pop--over oui-pop--over--arrow-bottom-right\" style=\"display: block; opacity: 1;\">\n    <li class=\"oui-pop--over__content\">\n      <div class=\"oui-pop--over__title\">Title</div>\n      <p>Ipsa officiis bad-news minus earum a aperiam! Aperiam reiciendis vitae nihil libero et, hic!</p>\n    </li>\n  </ul>\n</div>"},{"type":"html","description":null,"code":"<div class=\"height--100\">\n  <ul class=\"oui-pop--over oui-pop--over--arrow-bottom-center\" style=\"display: block; opacity: 1;\">\n    <li class=\"oui-pop--over__content\">\n      <div class=\"oui-pop--over__title\">Title</div>\n      <p>Ipsa officiis bad-news minus earum a aperiam! Aperiam reiciendis vitae nihil libero et, hic!</p>\n    </li>\n  </ul>\n</div>"},{"type":"html","description":null,"code":"<div class=\"height--100\">\n  <ul class=\"oui-pop--over oui-pop--over--arrow-bottom-left\" style=\"display: block; opacity: 1;\">\n    <li class=\"oui-pop--over__content\">\n      <div class=\"oui-pop--over__title\">Title</div>\n      <p>Ipsa officiis bad-news minus earum a aperiam! Aperiam reiciendis vitae nihil libero et, hic!</p>\n    </li>\n  </ul>\n</div>"},{"type":"html","description":null,"code":"<div class=\"height--100\">\n  <ul class=\"oui-pop--over oui-pop--over--arrow-left-bottom\" style=\"display: block; opacity: 1;\">\n    <li class=\"oui-pop--over__content\">\n      <div class=\"oui-pop--over__title\">Title</div>\n      <p>Ipsa officiis bad-news minus earum a aperiam! Aperiam reiciendis vitae nihil libero et, hic!</p>\n    </li>\n  </ul>\n</div>"},{"type":"html","description":null,"code":"<div class=\"height--100\">\n  <ul class=\"oui-pop--over oui-pop--over--arrow-left-center\" style=\"display: block; opacity: 1;\">\n    <li class=\"oui-pop--over__content\">\n      <div class=\"oui-pop--over__title\">Title</div>\n      <p>Ipsa officiis bad-news minus earum a aperiam! Aperiam reiciendis vitae nihil libero et, hic!</p>\n    </li>\n  </ul>\n</div>"},{"type":"html","description":null,"code":"<div class=\"height--100\">\n  <ul class=\"oui-pop--over oui-pop--over--arrow-left-top\" style=\"display: block; opacity: 1;\">\n    <li class=\"oui-pop--over__content\">\n      <div class=\"oui-pop--over__title\">Title</div>\n      <p>Ipsa officiis bad-news minus earum a aperiam! Aperiam reiciendis vitae nihil libero et, hic!</p>\n    </li>\n  </ul>\n</div>"}]};
  const reactData = {"description":"Simple component often used to display supplemental information to contents\non a page.\n\nSee `OverlayWrapper` component to position a popover on a page.","methods":[],"props":{"children":{"type":{"name":"node"},"required":true,"description":"Content that appears within the popover body"},"testSection":{"type":{"name":"string"},"required":false,"description":"Hook for automated JavaScript tests"},"title":{"type":{"name":"string"},"required":false,"description":"Text describing the popover contents"}},"private":false};

  let dataObject = {
    react: reactData, 
    examples: examples ? examples : null,
    readme: readme ? readme : null,
    sass: sassData,
  }
  
  class PopoverComponent extends React.Component {
  
    render() {
      return (
        <ComponentContent data={ dataObject } />
      );
    };
  }
  
  export default PopoverComponent
