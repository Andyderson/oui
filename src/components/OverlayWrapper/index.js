import React from 'react';
import Tether from 'tether';

/**
 * Intelligently position elements on a page.
 * @param {Object} props - Properties passed to component
 * @returns {ReactElement}
 */
class OverlayWrapper extends React.Component {
  constructor(props) {
    super(props);
    this.state = { isOverlayOpen: false };
  }

  componentDidMount() {
    let options = {
      element: this._overlayEl,
      target: this._activatorEl.firstChild,
      attachment: `${this.props.verticalAttachment} ${this.props.horizontalAttachment}`,
      constraints: [{
        to: 'window',
        attachment: 'together',
      }],
    };

    if (this.props.verticalTargetAttachment && this.props.horizontalTargetAttachment) {
      options.targetAttachment = `${this.props.verticalTargetAttachment} ${this.props.horizontalTargetAttachment}`;
    }

    this._tether = new Tether(options);
  }

  componentWillUnmount() {
    this._tether.destroy();
  }

  render() {
    const Children = React.Children.map(this.props.children, (child) => {
      return React.cloneElement(child, {
        onClick: (event) => {
          this.setState({
            'isOverlayOpen': !this.state.isOverlayOpen,
          });

          if (child.props.onClick) {
            child.props.onClick(event);
          }
        },
      });
    });

    return (
      /* eslint-disable react/jsx-no-bind */
      <div data-test-section={ this.props.testSection }>
        <div
          style={ { display: 'inline-block' } }
          ref={ (ref) => { this._activatorEl = ref; } }>
          { Children }
        </div>
        <div
          ref={ (ref) => { this._overlayEl = ref; } }
          style={ this.state.isOverlayOpen ? { display: 'block' } : { display: 'none' } }>
          { this.props.overlay }
        </div>
      </div>
      /* eslint-enable */
    );
  }
}

OverlayWrapper.propTypes = {
  /** Element that the `overlay` should attach to */
  children: React.PropTypes.node.isRequired,
  /** Side of the `overlay` that should attach to the `children` */
  horizontalAttachment: React.PropTypes.oneOf(['left', 'center', 'right']),
  /** Side of `children` that should attach to the `overlay` */
  horizontalTargetAttachment: React.PropTypes.oneOf(['left', 'center', 'right']),
  /** The element that is attached to the children */
  overlay: React.PropTypes.node.isRequired,
  /** Hook for automated JavaScript tests */
  testSection: React.PropTypes.string,
  /** Vertical edge of the `overlay` that should touch the `children` */
  verticalAttachment: React.PropTypes.oneOf(['top', 'middle', 'bottom']),
  /** Vertical edge of the `children` that should touch the `overlay` */
  verticalTargetAttachment: React.PropTypes.oneOf(['top', 'middle', 'bottom']),
};

OverlayWrapper.defaultProps = {
  horizontalAttachment: 'center',
  verticalAttachment: 'top',
};

export default OverlayWrapper;
