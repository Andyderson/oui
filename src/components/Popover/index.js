import React from 'react';

/**
 * Display information attached to a target element.
 *
 * See `OverlayTrigger` component to position a popover on a page.
 * @param {Object} props - Properties passed to component
 * @returns {ReactElement}
 */
const Popover = (props) => {
  return (
    <div
      className={ 'oui-pop--over' }
      style={ props.isVisible ? { display: 'block', opacity: 1 } : null }
      data-test-section={ props.testSection }>
      <div className="oui-pop--over__content">
        { props.title && (
          <div className="oui-pop--over__title">{ props.title }</div>
        ) }
        { props.children }
      </div>
    </div>
  );
};

Popover.propTypes = {
  /** Content that appears within the popover body */
  children: React.PropTypes.node.isRequired,
  /** Should the popover be visible on the page */
  isVisible: React.PropTypes.bool,
  /** Hook for automated JavaScript tests */
  testSection: React.PropTypes.string,
  /** Text describing the popover contents */
  title: React.PropTypes.string,
};

export default Popover;
