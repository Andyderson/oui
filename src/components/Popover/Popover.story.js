import React from 'react';

import { storiesOf, addDecorator } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import  { withKnobs, text, boolean, number, select } from '@storybook/addon-knobs';

import Popover from './index.js';
import OverlayWrapper from '../OverlayWrapper';
import Button from '../Button';

const stories = storiesOf('Popover', module);
stories
  .addDecorator(withKnobs)
  .addDecorator(story => (
    <div id="root-preview">
      {story()}
    </div>
  ));

stories
    .addWithInfo('default', () => <div>
        <div className="position--relative height--100 text--center">
        <Popover title="Lorem ipsum dolor sit amet">
            <p>
            Ipsa officiis bad-news minus earum a aperiam! Aperiam reiciendis vitae nihil libero et, hic!
            </p>
        </Popover>
        </div>
    </div>)
    .addWithInfo('popover in action', () => <div className="position--relative height--100 text--center">
        <OverlayWrapper
            behavior="click"
            horizontalAttachment="center"
            overlay={<Popover title="This is a working Popover!"><p>See `OverlayWrapper` for more examples.</p></Popover>}
            shouldHideOnClick
            verticalAttachment="top">
            
            <Button width="default"> Open Popover </Button>
        
        </OverlayWrapper>
    </div>)

