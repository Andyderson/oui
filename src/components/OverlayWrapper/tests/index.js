import React from 'react';
import OverlayWrapper from '../index';
import Popover from '../../Popover';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

describe('components/OverlayWrapper', () => {
  it('should render a Popover component inside', () => {
    const component = mount(
      <OverlayWrapper
        behavior="click"
        horizontalAttachment="left"
        horizontalTargetAttachment="left"
        overlay={ <Popover title="Lorem ipsum dolor sit amet">
          <p>Ipsa officiis bad-news minus earum a aperiam! Aperiam reiciendis vitae nihil libero et, hic!</p>
        </Popover> }
        shouldHideOnClick={ true }
        verticalAttachment="top"
        verticalTargetAttachment="bottom">
        <button>
          Open Popover
        </button>
      </OverlayWrapper>
    );
    expect(shallowToJson(component)).toMatchSnapshot();
  });
  it('should render component with `state.isOverlayOpen=true` when click if click bahavior', () => {
    const component = mount(
      <OverlayWrapper
        behavior="click"
        horizontalAttachment="left"
        horizontalTargetAttachment="left"
        overlay={ <Popover title="Lorem ipsum dolor sit amet">
          <p>Ipsa officiis bad-news minus earum a aperiam! Aperiam reiciendis vitae nihil libero et, hic!</p>
        </Popover> }
        shouldHideOnClick={ true }
        verticalAttachment="top"
        verticalTargetAttachment="bottom">
        <button>
          Open Popover
        </button>
      </OverlayWrapper>
    );
    component.find('.click-area').simulate('click');
    expect(component.state().isOverlayOpen).toBe(true);
  });
  it('should render component with `state.isOverlayOpen=true` when hover if hover behavior', () => {
    const component = mount(
      <OverlayWrapper
        behavior="hover"
        horizontalAttachment="left"
        horizontalTargetAttachment="left"
        overlay={ <Popover title="Lorem ipsum dolor sit amet">
          <p>Ipsa officiis bad-news minus earum a aperiam! Aperiam reiciendis vitae nihil libero et, hic!</p>
        </Popover> }
        shouldHideOnClick={ true }
        verticalAttachment="top"
        verticalTargetAttachment="bottom">
        <button>
          Open Popover
        </button>
      </OverlayWrapper>
    );
    component.find('.click-area').simulate('mouseover');
    expect(component.state().isOverlayOpen).toBe(true);
  });
  it('should render component with `state.isOverlayOpen=false` when clicking escape key', () => {
    const component = mount(
      <OverlayWrapper
        behavior="hover"
        horizontalAttachment="left"
        horizontalTargetAttachment="left"
        overlay={ <Popover title="Lorem ipsum dolor sit amet">
          <p>Ipsa officiis bad-news minus earum a aperiam! Aperiam reiciendis vitae nihil libero et, hic!</p>
        </Popover> }
        shouldHideOnClick={ true }
        verticalAttachment="top"
        verticalTargetAttachment="bottom">
        <button>
          Open Popover
        </button>
      </OverlayWrapper>
    );
    component.find('.click-area').simulate('click');
    component.simulate('keyDown', { keyCode: 40 });
    expect(component.state().isOverlayOpen).toBe(false);
  });
  it('should call prop function on click', () => {
    const component = mount(
      <OverlayWrapper
        behavior="hover"
        horizontalAttachment="left"
        horizontalTargetAttachment="left"
        overlay={ <Popover title="Lorem ipsum dolor sit amet">
          <p>Ipsa officiis bad-news minus earum a aperiam! Aperiam reiciendis vitae nihil libero et, hic!</p>
        </Popover> }
        shouldHideOnClick={ true }
        verticalAttachment="top"
        verticalTargetAttachment="bottom">
        <button>
          Open Popover
        </button>
      </OverlayWrapper>
    );
    component.find('.click-area').simulate('click');
    component.simulate('keyDown', { keyCode: 40 });
    expect(component.state().isOverlayOpen).toBe(false);
  });
  it('should instantiate OverlayWrapper', () => {
    const component = mount(
      <OverlayWrapper
        behavior="hover"
        horizontalAttachment="left"
        horizontalTargetAttachment="left"
        overlay={ <Popover title="Lorem ipsum dolor sit amet">
          <p>Ipsa officiis bad-news minus earum a aperiam! Aperiam reiciendis vitae nihil libero et, hic!</p>
        </Popover> }
        shouldHideOnClick={ true }
        verticalAttachment="top"
        verticalTargetAttachment="bottom">
        <button>
          Open Popover
        </button>
      </OverlayWrapper>
    );
    const inst = component.instance();
    expect(inst).toEqual.instanceOf(component);
  });
});
