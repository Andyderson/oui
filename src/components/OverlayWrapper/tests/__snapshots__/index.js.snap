// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/OverlayWrapper should render a Popover component inside 1`] = `
<OverlayWrapper
  behavior="click"
  horizontalAttachment="left"
  horizontalTargetAttachment="left"
  isConstrainedToScreen={false}
  overlay={
    <Popover
      title="Lorem ipsum dolor sit amet"
    >
      <p>
        Ipsa officiis bad-news minus earum a aperiam! Aperiam reiciendis vitae nihil libero et, hic!
      </p>
    </Popover>
  }
  shouldHideOnClick={true}
  verticalAttachment="top"
  verticalTargetAttachment="bottom"
>
  <div
    data-oui-component={true}
    style={
      Object {
        "display": "inline-block",
      }
    }
  >
    <div
      className="click-area"
      onClick={[Function]}
      style={
        Object {
          "display": "inline-block",
        }
      }
    >
      <button>
        Open Popover
      </button>
    </div>
    <div
      style={
        Object {
          "display": "none",
        }
      }
    >
      <Popover
        title="Lorem ipsum dolor sit amet"
      >
        <div
          className="oui-pop--over"
          data-oui-component={true}
          style={
            Object {
              "display": "block",
              "opacity": 1,
              "position": "initial",
            }
          }
        >
          <div
            className="oui-pop--over__content"
          >
            <div
              className="oui-pop--over__title"
            >
              Lorem ipsum dolor sit amet
            </div>
            <p>
              Ipsa officiis bad-news minus earum a aperiam! Aperiam reiciendis vitae nihil libero et, hic!
            </p>
          </div>
        </div>
      </Popover>
    </div>
  </div>
</OverlayWrapper>
`;
